SEPADirectDebit/src/main/java/com/braintreepayments/api/SEPADirectDebitClient.java

```diff
@@ -113,37 +86,57 @@ public void tokenize(final FragmentActivity activity,
                         } else {
                             braintreeClient.sendAnalyticsEvent(
                                     "sepa-direct-debit.create-mandate.failure");
-                            listener.onSEPADirectDebitFailure(
-                                    new BraintreeException("An unexpected error occurred."));
+                            callback.onResult(null, new BraintreeException("An unexpected error occurred."));
                         }
                     } else if (createMandateError != null) {
                         braintreeClient.sendAnalyticsEvent(
                                 "sepa-direct-debit.create-mandate.failure");
-                        listener.onSEPADirectDebitFailure(createMandateError);
+                        callback.onResult(null, createMandateError);
                     }
                 });
     }
 
-    void onBrowserSwitchResult(FragmentActivity activity) {
+    // TODO: - The wording in this docstring is confusing to me. Let's improve & align across all clients.
+    /**
+     * After receiving a result from the SEPA mandate web flow via
+     * {@link SEPADirectDebitLauncher#handleReturnToAppFromBrowser(Context, Intent)}, pass the
+     * {@link SEPADirectDebitBrowserSwitchResult} returned to this method to tokenize the SEPA
+     * account and receive a {@link SEPADirectDebitNonce} on success.
+     *
+     * @param sepaDirectDebitBrowserSwitchResult a {@link SEPADirectDebitBrowserSwitchResult} received
+     *                                           in the callback of {@link SEPADirectDebitLauncher}
+     * @param callback {@link SEPADirectDebitBrowserSwitchResultCallback}
+     */
+    public void onBrowserSwitchResult(@NonNull SEPADirectDebitBrowserSwitchResult sepaDirectDebitBrowserSwitchResult,
+                                      @NonNull final SEPADirectDebitBrowserSwitchResultCallback callback) {
         BrowserSwitchResult browserSwitchResult =
-                braintreeClient.deliverBrowserSwitchResult(activity);
+                sepaDirectDebitBrowserSwitchResult.getBrowserSwitchResult();
+        if (browserSwitchResult == null && sepaDirectDebitBrowserSwitchResult.getError() != null) {
+            callback.onResult(null, sepaDirectDebitBrowserSwitchResult.getError());
+            return;
+        }
+
+        if (browserSwitchResult == null) {
+            callback.onResult(null, new BraintreeException("An unexpected error occurred."));
+        }
 
         int result = browserSwitchResult.getStatus();
         switch (result) {
             case BrowserSwitchStatus.CANCELED:
                 braintreeClient.sendAnalyticsEvent("sepa-direct-debit.browser-switch.canceled");
-                listener.onSEPADirectDebitFailure(
-                        new UserCanceledException("User canceled SEPA Debit."));
+                callback.onResult(null, new UserCanceledException("User canceled SEPA Debit."));
                 break;
```

Comment:
This code block never executes on a user cancel clicking the "X" in the CCT (same for [this block](https://github.com/braintree/braintree_android/blob/v5/PayPal/src/main/java/com/braintreepayments/api/PayPalClient.java#L237-L241) in the PayPalClient).

I talked with @sarahkoop about this a little, and since the Launcher pattern in V5 is using the "manual browser switching" pattern from V4 she mentioned that we don't have a way to distinguish cancel cases. Currently for the cancel case, the [Launcher's BroswerSwitchClient.parseResult() call](https://github.com/braintree/braintree_android/blob/e94dec16aa0608c080f67bfe718658dac6363619/PayPal/src/main/java/com/braintreepayments/api/PayPalLauncher.java#L75) returns a null intent, which no-ops the method and doesn't notify the merchant that the flow is complete. This is currently the same for both PP and SEPA v5 launcher flows.

Wanted to loop in @sshropshire too. It sounds like this is expected for v5? Sarah mentioned a potential customization to CCT as a workaround.

Reply:
Yeah we still haven't been able to find a good way to signal a cancelation. CCT is super restrictive.